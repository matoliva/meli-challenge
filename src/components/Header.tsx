import logo from '../assets/logo.png'

interface IProps {
  children?: React.ReactNode
}
export const Header = ({children}: IProps) => {
  return (
    <header className="header" data-testid="header">
      <img className="header__logo" src={logo} alt="logo" />
      {children}
    </header>
  )
}
