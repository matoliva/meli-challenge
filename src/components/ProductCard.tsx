import noImage from '../assets/no-image.png'
import {Link} from 'react-router-dom'
import {IItemMapped} from '../pages/HomePage'

interface IProps {
  item: IItemMapped
}

export const ProductCard = ({item}: IProps) => {
  const {id, title, price, currency, picture, condition, address} = item

  const currencyMap = currency === 'ARS' ? '$' : 'uSd'
  const image = picture ? picture : noImage

  return (
    <Link to={`/item/${id}`} style={{textDecoration: 'none', color: 'black'}}>
      <article
        id={id}
        className="product-card pointer"
        data-testid="product-card"
      >
        <div className="product-card__image">
          <img src={image} alt="product" data-testid="product-card__image" />
        </div>
        <div className="product-card__description">
          <p
            className="product-card__description--price"
            data-testid="product-card__currency"
          >
            {currencyMap} {price}
          </p>
          <p data-testid="product-card__title">{title}</p>
          <p>{condition}</p>
        </div>
        <div className="product-card__location">
          <p data-testid="product-card__location">{address}</p>
        </div>
      </article>
    </Link>
  )
}
