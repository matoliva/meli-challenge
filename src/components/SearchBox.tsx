import logo from '../assets/logo.png'

export const SearchBox = () => {
  return (
    <div className="search-box">
      <img className="search-box__logo" src={logo} alt="logo" />

      <form className="search-box__form">
        <input
          type="text"
          className="input-text"
          placeholder=" Nunca dejes de buscar"
        />
        <div className="search-box__form__icon pointer"></div>
      </form>
    </div>
  )
}
