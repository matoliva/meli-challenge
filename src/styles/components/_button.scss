.btn {
  background-color: #3483fa;
  color: $secondary;
  font-size: medium;
  width: 100%;
  border: none;
  padding: 0.5rem 1rem 0.5rem 1rem;
  border-radius: 4px;

  &:hover {
    background-color: #2968c8;
  }
}
