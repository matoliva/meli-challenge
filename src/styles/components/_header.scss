.header {
  display: flex;
  gap: 1rem;
  width: 100%;
  align-items: center;
  background-color: $primary;
  padding: 0 5rem 0 5rem;
  height: 3rem;

  @media screen and (max-width: $mobile) {
    padding: 0 1rem 0 1rem;
  }

  &__logo {
    max-width: 3rem;
  }
}
