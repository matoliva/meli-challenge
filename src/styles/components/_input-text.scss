.input-text {
  height: 2rem;
  width: 100%;
  border-radius: 2px;
  border: none;
  outline-color: $light-grey;
}
