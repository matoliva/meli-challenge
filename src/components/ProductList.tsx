import {IItem} from '../models/products'
import {ProductCard} from './ProductCard'

interface IProp {
  items: IItem[]
}

export const ProductList = ({items}: IProp) => {
  return (
    <section data-testid="product-list">
      {items.map(item => (
        <ProductCard key={item.id} item={item} />
      ))}
    </section>
  )
}
