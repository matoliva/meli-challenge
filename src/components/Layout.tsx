import {Header} from './Header'
import {SearchBox} from './SearchBox'

interface IProps {
  children: React.ReactNode
}
export const Layout = ({children}: IProps) => {
  const onSearchChange = (searchText: string): void => {
    console.log(searchText)
  }
  return (
    <div className="main">
      <Header>
        <SearchBox onSearchChange={onSearchChange} />
      </Header>
      <main className="section">{children}</main>
    </div>
  )
}
