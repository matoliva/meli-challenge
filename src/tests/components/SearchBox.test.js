import {render, fireEvent} from '@testing-library/react'
import {SearchBox} from '../../components/SearchBox'

const baseProps = {
  onSearchChange: () => {},
}

describe('SearchBox', () => {
  beforeEach(() => {})

  it('renders correctly', () => {
    const {getByTestId} = render(<SearchBox {...baseProps} />)
    const searchBox = getByTestId('search-box')
    expect(searchBox).toMatchSnapshot()
  })

  it('should change the value when the user types', () => {
    const {getByTestId} = render(<SearchBox {...baseProps} />)
    const searchBoxInput = getByTestId('search-box__input')
    const textValue = 'new value'

    fireEvent.change(searchBoxInput, {target: {value: textValue}})

    expect(searchBoxInput.value).toBe(textValue)
  })

  it('sould call onSearchChange when the user hit the button', () => {
    const props = {
      ...baseProps,
      onSearchChange: jest.fn(),
    }

    const {getByTestId} = render(<SearchBox {...props} />)
    const textValue = 'new value'
    const searchBoxButton = getByTestId('search-box__button')
    const searchBoxInput = getByTestId('search-box__input')

    fireEvent.change(searchBoxInput, {target: {value: textValue}})
    fireEvent.click(searchBoxButton)

    expect(props.onSearchChange).toHaveBeenCalledTimes(1)
    expect(props.onSearchChange).toHaveBeenCalledWith(textValue)
    expect(searchBoxInput.value).toBe('')
  })

  it('sould call onSearchChange when the user press enter key', () => {
    const props = {
      ...baseProps,
      onSearchChange: jest.fn(),
    }

    const {getByTestId} = render(<SearchBox {...props} />)
    const textValue = 'new value'
    const searchBoxInput = getByTestId('search-box__input')

    fireEvent.change(searchBoxInput, {target: {value: textValue}})
    fireEvent.keyDown(searchBoxInput, {
      key: 'Enter',
      code: 'Enter',
      charCode: 13,
    })

    expect(props.onSearchChange).toHaveBeenCalledTimes(1)
    expect(props.onSearchChange).toHaveBeenCalledWith(textValue)
    expect(searchBoxInput.value).toBe('')
  })

  it('should not call onSearchChange when the textValue is empty', () => {
    const props = {
      ...baseProps,
      onSearchChange: jest.fn(),
    }

    const {getByTestId} = render(<SearchBox {...props} />)
    const searchBoxButton = getByTestId('search-box__button')
    const searchBoxInput = getByTestId('search-box__input')

    fireEvent.click(searchBoxButton)
    expect(props.onSearchChange).not.toHaveBeenCalled()

    fireEvent.keyDown(searchBoxInput, {
      key: 'Enter',
      code: 'Enter',
      charCode: 13,
    })
    expect(props.onSearchChange).not.toHaveBeenCalled()
  })
})
