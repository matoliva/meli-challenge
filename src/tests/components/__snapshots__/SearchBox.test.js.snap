// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox renders correctly 1`] = `
<form
  class="search-box__form"
  data-testid="search-box"
>
  <input
    class="input-text"
    data-testid="search-box__input"
    placeholder=" Nunca dejes de buscar"
    type="text"
    value=""
  />
  <button
    class="search-box__form__icon pointer"
    data-testid="search-box__button"
    type="submit"
  />
</form>
`;
